sockets/lib/multiplexer.go

```diff
@@ -0,0 +1,334 @@
+/**
+ * Multiplexer - Socket/Channel/Subchannel state machine
+ * https://pokemonshowdown.com/
+ *
+ * This keeps track of the sockets that connect to the SockJS server. Sockets
+ * are stored in the multiplexer to allow the parent process to manipulate them
+ * as it pleases. Channels represent rooms in the parent process; subchannels
+ * split battle rooms into three groups: side 1, side 2, and spectators.
+ * Certain messages will display differently depending on which subchannel the
+ * user's socket is in.
+ */
+
+package sockets
+
+import (
+	"fmt"
+	"net"
+	"path"
+	"regexp"
+	"strconv"
+	"sync"
+	"sync/atomic"
+
+	"github.com/igm/sockjs-go/sockjs"
+)
+
+// Subchannel IDs
+const (
+	DEFAULT_SUBCHANNEL_ID byte = '0'
+	P1_SUBCHANNEL_ID      byte = '1'
+	P2_SUBCHANNEL_ID      byte = '2'
+)
+
+// Map of socket IDs to subchannel IDs.
+type Channel map[string]byte
+
+type Multiplexer struct {
+	nsid     uint64                    // Socket ID counter.
+	sockets  map[string]sockjs.Session // Map of socket IDs to sockets.
+	smux     sync.Mutex                // nsid and sockets mutex.
+	channels map[string]Channel        // Map of channel (i.e. room) IDs to channels.
+	cmux     sync.Mutex                // channels mutex.
+	scre     *regexp.Regexp            // Regex for splitting subchannel broadcasts into their three messages.
```

Comment:
I think it fits better in the struct, since only `*Multiplexer.subchannelBroadcast` needs to be able to use it, unlike the other constants that exist